### DISCUSSION TITLE
Should we follow the principle of value for a property of Wikidata?

### DISCUSSION THREAD

**Discussion by candlecao** [DDMAL/linkedmusic-datalake#99] [General]
The principle is generally:
In Wikidata, almost a property is actually either linking another URI(object property) or linking just literal value(data property). 

For example, the value of property `wdt:P136` genre are all URI, as you can refer by query:
```
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT ?type (COUNT(?genre) AS ?count)
WHERE {
  ?item wdt:P136 ?genre .
  BIND(
    IF(
      isIRI(?genre),
      "URI",
      IF(
        isLiteral(?genre),
        "String",
        "Other"
      )
    ) AS ?type
  )
}
GROUP BY ?type
```
After executing this above, you see the count of URI is 1961941, but no even one literal data.

However, in our reconciliation task, we use string as the value of P136. Does that matter?

**Comment by fujinaga** [comment]
I don’t understand your question but have you looked into “music genre” Q188451?

**Reply by candlecao** [reply]
Thanks, Ich. I will look into it further.

**Comment by fujinaga** [comment]
Have you seen this https://www.wikidata.org/wiki/Wikidata:WikiProject_Music/Lists/Instance_of_music_genre?

**Comment by dchiller** [comment]
> However, in our reconciliation task, we use string as the value of P136. Does that matter?

So are you saying that after reconciliation, the object of triples with predicate P136 are all literals? For something like genre, where we know that there are genre items in wikidata, it seems to me like we didn't actually reconcile those values...

**Reply by ahankinson** [reply]
It’s not really “Linked Data” if there’s no link… 😁

**Reply by candlecao** [reply]
> > However, in our reconciliation task, we use string as the value of P136. Does that matter?
> 
> So are you saying that after reconciliation, the object of triples with predicate P136 are all literals? For something like genre, where we know that there are genre items in wikidata, it seems to me like we didn't actually reconcile those values...

I think I partially conveyed my meaning to you. To clarify:

> P136 is actually an "object property" in Wikidata.
> But we use P136 as a "data property" in our reconciliation.
> In our reconciled linked data, for genre, we use another property as object property instead, which is http://www.wikidata.org/prop/direct/P8052, called MusicBrainz genre ID.

Does that matter if we mix up object property and data property. I am not sure if this will interfere with the future project.

**Reply by dchiller** [reply]
In my mind, there should be a Wikidata URI as the object of P136 in our reconciled data. If there isn't, my contention is that we haven't reconciled the data against Wikidata. 

Am I understanding?

**Reply by candlecao** [reply]
Yes, I think you basically understood my meaning. But there is some trivial stuff to be explained.  Thanks.

**Comment by fujinaga** [comment]
Yes, there will be a Wikidata URI as the object for P136 unless the original genre URI is the Musicbrainz genre ID, in which case, we will use P8052 instead of P136.

**Reply by candlecao** [reply]
Actually the property value of P8052 is still literal instead of a URI. 
So we have temporarily determined to use P136 to link the genre URI based on the Musicbrainz genre ID. @Yueqiao12Zhang 
(And we can use reasoning tech to supplement data to let P136 link the corresponding wiki genre URI.)

Regardless, there is one principle we had better adhere to: if a property of wiki is an object property, we try persistently reusing it as an object property; if a property of wiki is a data property, we also persistently reuse it as a data property. Only when we cannot map to an appropriate property from Wikidata will we reuse properties elsewhere or invent properties for our own use.
--It's not a significant problem. We will document any particular situation that arises.

**Reply by dchiller** [reply]
> So we have temporarily determined to use P136 to link the genre URI based on the Musicbrainz genre ID. @Yueqiao12Zhang
> (And we can use reasoning tech to supplement data to let P136 link the corresponding wiki genre URI.)

I'm not sure why any reasoning would be required for this supplementation. MusicBrainz genres are reconciled with Wikidata... In other words, genres have both MusicBrainz ID's and Wikidata ID's and this linking has already occurred. Perhaps we're discussing the representation of that relationship in our Wikidata schema, but that relationship has already been determined, no?

**Reply by Yueqiao12Zhang** [reply]
The problem is that we have the Wikidata values for all entities in different CSV files. For example,
```
label_id    genre_id                   relations_wiki
123            /musicbrainz/genre/123        Q12345
456           /musicbrainz/genre/456        Q45678

genre_id                    relations_wiki
/musicbrainz/genre/123           Q123
/musicbrainz/genre/456           Q456
```

This causes a problem in the conversion process. We must go through this path:
label_id    >>P8052 or P136>> genre_id >>P2888>> relations_wiki

**Reply by dchiller** [reply]
I don't understand. 

Let's say I have a MusicBrainz genre [pop rock](https://musicbrainz.org/genre/797e2e85-5ffd-495c-a757-8b4079052f0e). 

This has a MusicBrainz ID (797e2e85-5ffd-495c-a757-8b4079052f0e). MusicBrainz has already reconciled this genre with Wikidata to have a Wikidata QID ([Q484641](https://www.wikidata.org/wiki/Q484641)). In your recent PR (#80), you've developed a process to obtain this QID.

I also have the "release-group" of [Fleetwood Mac's Rumors](https://musicbrainz.org/release-group/416bb5e5-c7d1-3977-8fd7-7c9daf6c2be6) which has the genre "pop rock".

```mermaid
  graph TD;
      A[Rumors release-group]--P136-->C[https://www.wikidata.org/wiki/Q484641];
      C--P8052--->D[797e2e85-5ffd-495c-a757-8b4079052f0e];
```

This seems doable and seems to properly conform to object vs data distinction.

**Reply by Yueqiao12Zhang** [reply]
Using this convention, since in my schema, I link the graphs using MBIDS but not Wikidata IDs, we must to through
```mermaid
  graph TD;
      A[Rumors release-group]--P8052--->C[797e2e85-5ffd-495c-a757-8b4079052f0e];
      C--P136-->D[https://www.wikidata.org/wiki/Q484641];
```
Since we have to track using MBID before we can access the QID for that object.

**Reply by dchiller** [reply]
Ok, but I don't think that's what P8052 means. It looks to me like the subject of P8052 is a genre and the object is a literal MusicBrainz ID.

**Reply by dchiller** [reply]
PS. I edited your comment so the mermaid graph would display.

**Reply by Yueqiao12Zhang** [reply]
Yes, that's where we were having problems because we were deciding to use P136 or P8052. Now we think we should remove the genre_name column because we don't need it. The problem is which property we should use when linking a subject to genre MBID.

**Reply by dchiller** [reply]
> The problem is which property we should use when linking a subject to genre MBID.

A release group, for example, will have a genre. That genre may/will have multiple reconciled identifiers (it'll have a Wikidata ID, it'll have a MusicBrainz ID, maybe one day it'll have a "Dylan's great big ontology" ID) ... but I don't think there's any question of which of P136 or P8052 to use to link a release group (or a recording, or a song, etc) with a genre! It's P136!

**Comment by Yueqiao12Zhang** [comment]
So that's what we decided to do:
```mermaid
  graph TD;
      A[https://musicbrainz.org/release-group/416bb5e5-c7d1-3977-8fd7-7c9daf6c2be6]--P136--->C[https://musicbrainz.org/genre/797e2e85-5ffd-495c-a757-8b4079052f0e];
      C--P2888-->D[https://www.wikidata.org/wiki/Q484641];
```
Then we use the reasoner in Virtuoso to complete the relation:
```mermaid
  graph TD;
      A[https://musicbrainz.org/release-group/416bb5e5-c7d1-3977-8fd7-7c9daf6c2be6]--P136--->D[https://www.wikidata.org/wiki/Q484641];
```

**Reply by dchiller** [reply]
FYI, use "```mermaid" to get the graph to display.

**Reply by dchiller** [reply]
So what is in our graph? The first, the second, or both?

**Reply by Yueqiao12Zhang** [reply]
My code makes the first, and in Virtuoso, Junjun converts it to the second.

**Comment by fujinaga** [comment]
The reason I mentioned P8052 is because I wanted to ensure that there's an indication that the genre was assigned by MusicBrainz. I'm not sure what's the best way to do this. This is to differentiate from some other music database that uses MusicBrainz genre ID for their assigned genre.

**Reply by candlecao** [reply]
OK, got it. We will handle this issue more clearly.
We may talk about it face to face.
